### ISSUE TITLE
Url encoding of backtick ("`") character in query strings

### ISSUE THREAD

**Issue by ethanresnick** [visionmedia/superagent#1367]
I have a query string that includes a backtick character in it, and the server I'm communicating with expects this character to not be percent encoded. However, superagent encodes it. Superagent's behavior is correct according to RFC 3986, but incorrect according to the WHATWG URL spec, which tracks what browsers do. My guess is that it'd be generally more useful to follow the browser behavior, but I'd at least like to see an option to be able to customize the behavior to match the browser; right now, there doesn't seem to be any way for me to use superagent to make the request with the unencoded character.

I'd note that superagent already follows browser behavior wrt percent encoding the single quote character, which RFC 3986 says shouldn't be percent encoded. This means it's also impossible to make a request with the single quote unencoded through superagent, even though that would be RFC 3986-compliant. Not being able to use either the backtick or the single quote makes it really hard to build nice DSLs for use in query strings.

**Comment by ethanresnick** [comment]
To be clear, this happens even when I call `request.query(String)` with my unencoded string, and a bit of debugging reveals that the issue goes all the way down to node's built-in `url.parse` (i.e., ``url.parse("http://example.com/?test=`")`` returns `{ ..., query: "test=%60" }`).

Node's solution to this issue is to offer both the RFC 3986-compliant `url.parse` and the WHATWG URL API in core, so perhaps superagent should simply surface both of those options? Note that using the built-in `new (require('url').URL)("http://example.com/?test=`").search` does leave the backtick unescaped.

**Comment by kornelski** [comment]
We depend on ``encodeURIComponent("`")`` which returns `'%60'`. 

It's not clear to me whether WHATWG URL recommends unescaped ``"`"`` or merely tolerates it as a recoverable error. The [path encode set](https://url.spec.whatwg.org/#path-percent-encode-set) contains [fragment set](https://url.spec.whatwg.org/#fragment-percent-encode-set), which contains backtick.

In superagent you can set query as a string ``req.query("foo=`")``, and IIRC it won't escape it in the Node version (in the browser version it depends on what the browser enforces).

But in general I've got a feeling your server depends on a problematic feature. Given decades-old mess of fragmented specs for URL/URI/IRI the server should be prepared to support the lowest common denominator.

**Comment by ethanresnick** [comment]
> We depend on encodeURIComponent("`") which returns '%60'. We still support IE, so we can't easily switch to the URL API.... In superagent you can set query as a string req.query("foo=`"), and IIRC it won't escape it in the Node version (in the browser version it depends on what the browser enforces).

The issue here isn't `encodeURIComponent`; even in the Node version of superagent with ``req.query("foo=`")``, the backtick gets escaped, because of superagent's internal use of `url.parse`. I'm pretty sure [this](https://github.com/visionmedia/superagent/blob/2eed60fd38fe0b61cd476fa157e05ce1178459b5/lib/node/index.js#L593) is the problematic call/the spot where node's WHATWG API could be used instead.

> It's not clear to me whether WHATWG URL recommends unescaped "`" or merely tolerates it as a recoverable error. The path encode set contains fragment set, which contains backtick.

I agree that the spec's pretty hard to follow. I don't think the path encode set is relevant, though, as that seems to only be referenced in the path processing steps, not the query ones. My best reading is that the key bit is the [query state](https://url.spec.whatwg.org/#query-state) of the parser, where the spec explicitly doesn't include backtick as one of the characters that gets percent encoded during parsing. (Normally, I'd expect encoding to be part of the serialization step, but that's not how the spec is structured; the serialization is just dumb concatenation, with the percent encoding having already happened during parsing/construction. Either way, the point is that, under WHATWG URL, ``serialize(parse(http://exmaple.com?t=`))`` leaves the backtick as is.)

> But in general I've got a feeling your server depends on a problematic feature. Given decades-old mess of fragmented specs for URL/URI/IRI the server should be prepared to support the lowest common denominator.

I kind of agree with this, and I'm probably going to switch the server to a new delimiter. However, let me give a bit more context. In my use case, I need a way to delimit literal strings from other query param data, and I think a reasonable approach would be to allow the client to delimit the string with _either_ literal single quotes (for RFC 3986-compliant clients, which should leave these unencoded) or by backticks (for WHATWG based clients, like browsers, that will encode the single quotes but not the backticks). However, even this scheme of allowing either delimiter doesn't work with superagent, because it unconditionally encodes both characters. To me, that seems pretty bad/frustrating. 

I think it's reasonable for new client libraries to correctly support *at least one of* RFC 3986 or WHATWG URL. Superagent, by encoding *both* backtick and single quote, is supporting neither. This isn't really it's fault — it all goes back to `url.parse` encoding single quotes in violation of RFC 3986, but that's probably good for security (that's why browsers do it), so I wouldn't suggest changing that in Node core even if it weren't a big breaking change. The upshot of all this, though, is that it's still impossible to make a bunch of useful requests from superagent. That's what I'd like to see some escape hatch for.

**Comment by kornelski** [comment]
You're effectively making quotes a new special character, and creating your own new syntax for the query string. Consider using something else, e.g. double URL-encode or your own escaping within query string values. Maybe use `\t` as a delimiter? Or if you're adventurous, `NUL` (`%00`).

**Comment by ethanresnick** [comment]
> You're effectively making quotes a new special character, and creating your own new syntax for the query string.

Not sure I follow; single quotes already *are* a special character under RFC 3986 — they're part of the sub-delims set, which is for unencoded use in query strings precisely to enable custom DSL syntaxes like this. That's my point...

**Comment by kornelski** [comment]
It's not *impossible*. You can have any query syntax you like with:

```js
req.get("http://foo").query("``````````")
```

This sends `'GET /?`````````` HTTP/1.1\r\nHost: foo\r\nAccept-Encoding: gzip, deflate\r\nCookie: \r\nConnection: close\r\n\r\n'`

**Comment by kornelski** [comment]
I've checked superagent's code. As far as superagent's own code goes, and it carefully preserves encoding of both `'` and `` ` `` when you use `.query("")`. It all breaks due to Node's `url.parse()`. Node's `http` module requires parsed URL, so that's hard to avoid. The WHATWG parser was added in 7.0, so we won't be able to use it until April 2019. I'm reluctant to add another 3rd party parser just to support such an edge case.

**Comment by ethanresnick** [comment]
> Node's `http` module requires parsed URL, so that's hard to avoid. 

Ahh, I didn't know this. Definitely makes things trickier.

> The WHATWG parser was added in 7.0, so we won't be able to use it until April 2019 (6.0 EOL). I'm reluctant to add another 3rd party parser just to support such an edge case.

Fair enough. Would it be possible to do some very minimal post-processing on the result of `url.parse` to add back the unescaped versions of any `'` or ``` that were present in the `.query()` arg?

**Comment by kornelski** [comment]
I've used ``` `` ` `` ```

**Comment by kornelski** [comment]
Is it allowed to have unescaped `'`/`` ` `` in the path? ``path.replace(/%60/g,'`')`` may be an OK solution,

**Comment by ethanresnick** [comment]
Yeah, the above obviously isn't as simple as a one-liner; that said, it doesn't seem too bad to me. Is there something in particular that bothers you about it?

Also, just to clarify, the above snippet could be used across the board, without superagent importing `require("url").URL` at all. (That is, I didn't mean to suggest importing the WHATWG stuff conditionally and then falling back to the above; that would be needless complexity and open the door to bugs caused by WHATWG URL doing other subtle things that the non-WHATWG-superagent still wouldn't do.)

If the above snippet were implemented, it could get applied to all requests unconditionally (i.e., no need for a user facing option either) because applying it unconditionally doesn't actually take away any expressive power; a user who wants the escaping can still get it (by escaping the `` ` `` before passing it in). Also, in practice, applying it automatically is likely to actually increase interop, as it's what browsers do, and I imagine pretty much every server should be able to accept requests from browsers.

That said, applying this automatically would technically be a breaking change, so it probably should live behind a user opt-in option just until the next major version.

**Comment by ethanresnick** [comment]
@kornelski Any more thoughts on this?

**Comment by kornelski** [comment]
It seems that `url.parse()` is guilty here, so it should be changed. You should probably raise this issue with the Node core, but since it's a stable API they probably won't like the idea of changing it.

I'd be OK with moving to the WHATWG parser, except for the pain of Node 6 compatibility. I'm worried that conditional `require()` will have nasty side effects in bundlers (we already get super weird bug reports about them).

The workaround you proposed is narrow (requires query as a string) and little bit on the convoluted side. It feels like tech debt for a solution that you will be the only person in the world to ever use.

**Comment by kornelski** [comment]
OK, if you maintain it, I can accept the change :)

**Comment by ethanresnick** [comment]
It's a deal :) I'll work up a PR tomorrow or Friday.

After that, how do you want to me to handle maintenance? I imagine you'd just tag me if/when any relevant bugs come up?

Also, for the PR, how do you want to land it: as a major version bump, or by adding an opt-in option to `request.query`? Unless there's some big cost to releasing a new major, I'd strongly prefer to do that; an option would have no good reason to exist long-term and would be a big API kludge. (As I mentioned earlier, if a user wants the old behavior, they can get it with ``request.query(oldInput.replace(/`/g, '%60'))`` — no option needed.)

**Comment by kornelski** [comment]
I've just bumped major, which I plan to release on April 30th

**Comment by ethanresnick** [comment]
Cool. So, I imagine that means that, if I get you the PR ASAP, it can go out with that release?

**Comment by kornelski** [comment]
Yes

**Comment by ethanresnick** [comment]
Closed by #1369.
